### DISCUSSION TITLE
Finding the control with focus

### DISCUSSION THREAD

**Discussion by dasien** [tui-cs/Terminal.Gui#3328] [Q&A] [answered]
Hello,

I just want to check my findings on something. I am using v.1.15.1.  In the `Enter` event of a control, say a `TextField`, i want to select all the text in that control, so that tabbing between `TextFields` will highlight all the text contained within.  If that `TextField` is embedded in a `ScrollView`, which is itself embedded in a `View`, do i need to build an involution to walk the tree of `SubViews` of the outer `View' then the `ScrollView`, checking along the way that the `ScrollView.HasFocus = true` (in case some other `SubView' of the outer `View' got focus) and then finding the control inside the `ScrollView.SubViews` to get to the `TextField` that also `HasFocus = true`?

Then, with that reference, i can `SelectAll` on the `Text` of that control?  

I wonder if it would be possible as part of the `FocusEventArgs` a reference to the control on which the event was set could also be passed, so that in addition to the control which lost focus, the control which is getting focus would also be known.

Am missing a faster/better way to get the reference to the control receiving focus?

Thanks!

**Comment by BDisp** [comment]
The `Enter` and the `Leave` event pass in the parameter argument the view that is losing/get focus, I think. To get it visible in the `ScrollView` you have to check the bounds location and scroll to that position. That is a lot of work but doable, I think.

**Comment by dasien** [comment]
Thank you for the reply.  I am not trying to make the control visible in the `ScrollView`, that was just an example of a container into which other `Views` would be placed.  I am really just trying to get a reference to the control that got focus, so that if it is a `TextField`, i can then execute the `SelectAllText` method on it.  Other than subclassing the `TextField` so that i can override `OnEnter` method, i am not sure of another way to determine which control got focus when it is embedded in other views, other than the loop i was suggesting where i iterate over the parent `View` with `HasFocus` set to `true` and then look through its `SubViews` to see which of those has focus until i arrive at a view which both has focus and has no more subviews to check.  

I was just making sure there wasn't an easier way to determine the control which got focus.

**Reply by BDisp** [reply]
You can subscribe the `Enter` event of the `TextField` for that. When you tab between the views and the `TextField` get focus the `Enter` will be raised and if write a code like, `((TextField)sender).SelectAllText()`, it will select all the text.

**Comment by dasien** [comment]
Thank you very much.  I am currently subscribed to that Enter event on the `TextField` in my application.  That was actually how i started, but i don't see sender as a property of the `FocusEventArgs` that are passed in as part of that event.  `sender` doesn't seem to be part of the event signature.  How to i get to that `sender` to cast it?

**Reply by BDisp** [reply, marked as answer]
In the v1 the events only use the `event` and not the `event EventHandler` and I forgot that. What you can do is in the event call a method that pass the sender to it. e.g.

```cs
var tfUserName = new TextField();
tfUserName.Enter += (_) => OnEnter(tfUserName)

var tfUserPwd = new TextField();
tfUserPwd.Enter += (_) => OnEnter(tfUserPwd)

private void OnEnter(View sender)
{
   ((TextField)sender).SelectAllText();
}
```

**Reply by dasien** [reply]
Thank you very much!
